File: src/actool_linux/multi_database.py
"""Multiple BOM database support for legacy CoreUI compatibility.

This module provides support for reading and writing CAR files with
multiple specialized BOM databases, as used in older CoreUI versions.
"""

from typing import Dict, List, Optional, Tuple
from pathlib import Path
from .bom import BOMStore, BOMError
import struct


class MultiDatabaseCAR:
    """CAR file with multiple BOM databases.
    
    Older CoreUI versions (< 700) use a single BOMStore, while newer
    versions (700+) may use multiple specialized databases:
    - imagedb: Image renditions
    - colordb: Color definitions
    - fontdb: Font definitions
    - fontsizedb: Font size definitions
    - appearancedb: Appearance definitions
    - facetKeysdb: Facet keys (theme identifiers)
    - bitmapKeydb: Bitmap keys
    - zcbezeldb: Zero-code bezel database
    - zcglyphdb: Zero-code glyph database
    """
    
    def __init__(self, coreui_version: int = 975):
        self.coreui_version = coreui_version
        self.databases: Dict[str, BOMStore] = {}
        self.main_store: Optional[BOMStore] = None
        
        # Determine which databases to use based on version
        if coreui_version >= 700:
            self.database_names = [
                'imagedb', 'colordb', 'fontdb', 'fontsizedb',
                'appearancedb', 'facetKeysdb', 'bitmapKeydb'
            ]
            if coreui_version >= 850:
                self.database_names.extend(['zcbezeldb', 'zcglyphdb'])
        else:
            # Single database for old versions
            self.database_names = []
    
    @classmethod
    def from_path(cls, path: Path, coreui_version: int = 975) -> 'MultiDatabaseCAR':
        """Load a CAR file with multiple databases."""
        car = cls(coreui_version)
        
        # Load main BOMStore
        car.main_store = BOMStore.from_path(path)
        
        # Check for additional databases
        for db_name in car.database_names:
            try:
                # Try to load as separate database
                db_path = path.parent / f"{path.stem}_{db_name}{path.suffix}"
                if db_path.exists():
                    car.databases[db_name] = BOMStore.from_path(db_path)
            except Exception:
                # Database not found or invalid, skip
                pass
        
        return car
    
    def get_database(self, name: str) -> Optional[BOMStore]:
        """Get a specific database by name."""
        return self.databases.get(name)
    
    def has_database(self, name: str) -> bool:
        """Check if a database exists."""
        return name in self.databases
    
    def get_all_databases(self) -> Dict[str, BOMStore]:
        """Get all databases."""
        return self.databases.copy()
    
    def get_image_renditions(self) -> List[dict]:
        """Get all image renditions from imagedb or main store."""
        if self.has_database('imagedb'):
            store = self.databases['imagedb']
        else:
            store = self.main_store
        
        if store is None:
            return []
        
        # Extract renditions from RENDITIONS tree
        renditions = []
        try:
            if 'RENDITIONS' in store.variables:
                # Parse renditions (simplified)
                renditions_data = store.named_block('RENDITIONS')
                # TODO: Parse actual rendition data
                pass
        except BOMError:
            pass
        
        return renditions
    
    def get_color_definitions(self) -> Dict[str, dict]:
        """Get all color definitions from colordb or main store."""
        if self.has_database('colordb'):
            store = self.databases['colordb']
        else:
            store = self.main_store
        
        if store is None:
            return {}
        
        colors = {}
        try:
            if 'COLORDEFINITIONS' in store.variables:
                # Parse color definitions (simplified)
                color_data = store.named_block('COLORDEFINITIONS')
                # TODO: Parse actual color data
                pass
        except BOMError:
            pass
        
        return colors
    
    def get_facet_keys(self) -> Dict[str, int]:
        """Get all facet keys from facetKeysdb or main store."""
        if self.has_database('facetKeysdb'):
            store = self.databases['facetKeysdb']
        else:
            store = self.main_store
        
        if store is None:
            return {}
        
        facet_keys = {}
        try:
            if 'FACETKEYS' in store.variables:
                # Parse facet keys (simplified)
                facet_data = store.named_block('FACETKEYS')
                # TODO: Parse actual facet key data
                pass
        except BOMError:
            pass
        
        return facet_keys
    
    def write_multi_database_car(self, output_path: Path, renditions: List[dict], 
                                 colors: Dict[str, dict], facet_keys: Dict[str, int]):
        """Write a CAR file with multiple databases.
        
        For CoreUI >= 700, creates separate database files.
        For CoreUI < 700, creates a single BOMStore.
        """
        if self.coreui_version >= 700 and len(self.database_names) > 0:
            # Write separate databases
            for db_name in self.database_names:
                db_path = output_path.parent / f"{output_path.stem}_{db_name}{output_path.suffix}"
                # TODO: Implement actual database writing
                pass
        else:
            # Write single database
            # TODO: Implement single database writing
            pass
    
    def validate_compatibility(self) -> Tuple[bool, str]:
        """Validate that the CAR file is compatible with the CoreUI version."""
        # Check database requirements
        if self.coreui_version >= 700:
            required_dbs = ['imagedb', 'colordb', 'facetKeysdb']
            for db_name in required_dbs:
                if not self.has_database(db_name) and self.main_store is None:
                    return False, f"Missing required database: {db_name}"
        
        if self.coreui_version >= 850:
            optional_dbs = ['zcbezeldb', 'zcglyphdb']
            # These are optional, just warn if missing
            for db_name in optional_dbs:
                if not self.has_database(db_name):
                    pass  # Optional, no error
        
        return True, ""

File: src/core/multi_database.rs
use crate::bom::BOMStore;
use std::collections::HashMap;
use std::path::Path;

pub struct MultiDatabaseCAR {
    pub coreui_version: u32,
    pub databases: HashMap<String, BOMStore>,
    pub main_store: Option<BOMStore>,
    pub database_names: Vec<String>,
}

impl MultiDatabaseCAR {
    pub fn new(coreui_version: u32) -> Self {
        let mut db_names = Vec::new();
        if coreui_version >= 700 {
            db_names.extend(vec![
                "imagedb".to_string(),
                "colordb".to_string(),
                "fontdb".to_string(),
                "fontsizedb".to_string(),
                "appearancedb".to_string(),
                "facetKeysdb".to_string(),
                "bitmapKeydb".to_string(),
            ]);
            if coreui_version >= 850 {
                db_names.push("zcbezeldb".to_string());
                db_names.push("zcglyphdb".to_string());
            }
        }

        Self {
            coreui_version,
            databases: HashMap::new(),
            main_store: None,
            database_names: db_names,
        }
    }

    pub fn from_path<P: AsRef<Path>>(path: P, coreui_version: u32) -> Result<Self, String> {
        let store = BOMStore::from_path(&path).map_err(|e| e.to_string())?;
        let mut car = Self::new(coreui_version);

        for name in &car.database_names {
            if let Some(parent) = path.as_ref().parent() {
                let stem = path.as_ref().file_stem().unwrap_or_default().to_string_lossy();
                let ext = path.as_ref().extension().unwrap_or_default().to_string_lossy();
                let db_path = parent.join(format!("{}_{}.{}", stem, name, ext));
                if db_path.exists() {
                    if let Ok(db_store) = BOMStore::from_path(db_path) {
                        car.databases.insert(name.clone(), db_store);
                    }
                }
            }
        }

        car.main_store = Some(store);
        Ok(car)
    }

    pub fn get_database(&self, name: &str) -> Option<&BOMStore> {
        self.databases.get(name)
    }

    pub fn has_database(&self, name: &str) -> bool {
        self.databases.contains_key(name)
    }

    pub fn get_all_databases(&self) -> &HashMap<String, BOMStore> {
        &self.databases
    }

    pub fn get_image_renditions(&self) -> Vec<Vec<u8>> {
        let mut rends = Vec::new();
        let target_store = self.get_database("imagedb").or(self.main_store.as_ref());
        if let Some(store) = target_store {
            if let Ok(data) = store.named_block_data("RENDITIONS") {
                rends.push(data.to_vec());
            }
        }
        rends
    }

    pub fn get_color_definitions(&self) -> HashMap<String, Vec<u8>> {
        let mut colors = HashMap::new();
        let target_store = self.get_database("colordb").or(self.main_store.as_ref());
        if let Some(store) = target_store {
            if let Ok(data) = store.named_block_data("COLORDEFINITIONS") {
                colors.insert("colordb".to_string(), data.to_vec());
            }
        }
        colors
    }

    pub fn get_facet_keys(&self) -> HashMap<String, u32> {
        let mut keys = HashMap::new();
        let target_store = self.get_database("facetKeysdb").or(self.main_store.as_ref());
        if let Some(store) = target_store {
            for (k, &v) in &store.variables {
                keys.insert(k.clone(), v);
            }
        }
        keys
    }

    pub fn write_multi_database_car<P: AsRef<Path>>(&self, _path: P) -> Result<(), String> {
        Ok(())
    }

    pub fn validate_compatibility(&self) -> (bool, String) {
        if self.coreui_version >= 700 {
            for req in &["imagedb", "colordb", "facetKeysdb"] {
                if !self.has_database(req) && self.main_store.is_none() {
                    return (false, format!("Missing required database: {}", req));
                }
            }
        }
        (true, String::new())
    }
}
